File: Haskell/HelloWorld/src/Extra/Road.hs
{-# LANGUAGE NamedFieldPuns #-}
module Extra.Road where

import Prelude

import Data.Foldable (foldl')

data Route = Route
  { destination :: Int
  , path :: [(Char, Int)]
  } deriving (Eq, Ord, Show)

data RoutingResult = RoutingResult
  { a :: Route
  , b :: Route
  } deriving (Eq, Ord, Show)

file :: String
file = "road.txt"

parseMap :: String -> [Int]
parseMap s = readInt <$> concat (words <$> lines s)

readInt :: String -> Int
readInt s = read s :: Int

groupValues :: [c] -> [(c, c, c)] -> [(c, c, c)]
groupValues [] acc = reverse acc
groupValues (a : b : x : xs) acc = groupValues xs ((a, b, x) : acc)
groupValues _ acc = reverse acc

shortestSteps :: RoutingResult -> (Int, Int, Int) -> RoutingResult
shortestSteps RoutingResult
  { a = Route
    { destination = destA
    , path = pathA
    }
  , b = Route
    { destination = destB
    , path = pathB
    }
  } (a, b, x) =
    let optA1 = Route { destination = destA + a, path = ('a', a) : pathA }
        optA2 = Route { destination = destB + b + x, path = ('x', x) : ('b', b) : pathB }
        optB1 = Route { destination = destB + b, path = ('b', b) : pathB }
        optB2 = Route { destination = destA + a + x, path = ('x', x) : ('a', a) : pathA } in
    RoutingResult
      { a = min optA1 optA2
      , b = min optB1 optB2
      }

initialRoutingResult :: RoutingResult
initialRoutingResult = RoutingResult
  { a = Route
    { destination = 0
    , path = []
    }
  , b = Route
    { destination = 0
    , path = []
    }
  }

optimalPath :: Foldable t => t (Int, Int, Int) -> [(Char, Int)]
optimalPath values =
  let RoutingResult { a, b } = foldl' shortestSteps initialRoutingResult values in
  let Route { path = resolved }
        | head (path a) /= ('x', 0) = a
        | head (path b) /= ('x', 0) = b
        | otherwise = Route { destination = 0, path = [] } in
  reverse resolved

run :: IO [(Char, Int)]
run = do
  rawString <- readFile file
  let values = flip groupValues [] $ parseMap rawString
  pure $ optimalPath values

File: FSharp/HelloWorld/Lib/Extra/Road.fs
module HelloWorld.Lib.Extra.Road

open System
open System.IO
open System.Text.RegularExpressions

type Route = {
    Destination: int
    Path: (char * int) list
}

type RoutingResult = {
    A: Route
    B: Route
}

let rec private GroupValues acc input =
    match input with
    | [] -> List.rev acc
    | a :: b :: x :: xs -> GroupValues ((a, b, x) :: acc) xs
    | _ -> List.rev acc

let private ParseMap (s: string) =
    let regex = Regex(@"\s")
    regex.Split(s)
    |> List.ofArray
    |> List.filter (fun elem -> not (String.IsNullOrWhiteSpace(elem)))
    |> List.map Int32.Parse
    
let private ShortestSteps (acc: RoutingResult) elem =
    let a, b, x = elem
    let optA1 = { Route.Destination = acc.A.Destination + a; Route.Path = ('a', a) :: acc.A.Path }
    let optA2 = { Route.Destination = acc.B.Destination + b + x; Route.Path = ('x', x) :: ('b', b) :: acc.B.Path }
    let optB1 = { Route.Destination = acc.B.Destination + b; Route.Path = ('b', b) :: acc.B.Path }
    let optB2 = { Route.Destination = acc.A.Destination + a + x; Route.Path = ('x', x) :: ('a', a) :: acc.A.Path }
    { RoutingResult.A = min optA1 optA2; RoutingResult.B = min optB1 optB2 }
    
let private OptimalPath values =
    let initialRoutingResult = { RoutingResult.A = { Route.Destination = 0; Route.Path = [] }; RoutingResult.B = { Route.Destination = 0; Route.Path = [] } }
    let routingResult = List.fold ShortestSteps initialRoutingResult values
    let route = if (List.head routingResult.A.Path) <> ('x', 0) then routingResult.A
                else if (List.head routingResult.B.Path) <> ('x', 0) then routingResult.B
                else { Route.Destination = 0; Route.Path = [] }
    List.rev route.Path
    
let Run =
    File.ReadAllText("road.txt")
    |> ParseMap
    |> GroupValues []
    |> OptimalPath